compiler-core/src/javascript/tests/snapshots/gleam_core__javascript__tests__bools__shadowed_bools_and_nil_typescript.snap

```diff
@@ -4,11 +4,17 @@ expression: "\npub type True { True False Nil }\npub fn go(x, y) {\n  let assert
 ---
 import type * as _ from "../gleam.d.mts";
 
-export class True extends _.CustomType {}
+export class True extends _.CustomType {
+  private __gleam__my__mod__True: never;
```

Comment:
These values do not have this field, so this would be incorrect. Is there a way to make these types nominal without making them incorrect?

Reply:
Hmm yes, this is sort of the point though, i.e. the desired type safety is then enforced by TypeScript _without_ the runtime overhead of actually creating these unused fields for real.

Making the types `private` and of type `never` is intended to stop anyone ever trying to access them. They'd get `undefined` were they to do so. Maybe a type of `undefined` instead of `never` would therefore be better? Or a name that more explicitly indicates not to access the field?

I'll have a think about whether there's an alternative approach here that doesn't involve creating these fields on the underlying object.

Reply:
It's not the value of the field, it's that there is a field at all. There is no field with that name on that type.

Why it is that TypeScript considers them the same? It woudn't be the same in TypeScript at runtime as they have different constructors and so would not pass instanceof checks.

Reply:
100% there's no field with that name on the type. Here's a hopefully clearer illustration of the problem that's trying to be addressed by this change.

If I write this Gleam code:

```gleam
pub type Thing {
  Thing
}
```

Its TypeScript will be:

```ts
import type * as _ from "../gleam.d.mts";

export class Thing extends _.CustomType {}

export type Thing$ = Thing;
```

Note also that Gleam's `option.Option` type compiles to this TypeScript:

```ts
export class Some<I> extends _.CustomType {
  constructor(argument$0: I);
  0: I;
}

export class None extends _.CustomType {}

export type Option$<I> = Some<I> | None;
```

Now, because `option.None` and `thing.Thing` are structurally identical, they're interchangeable in TypeScript. The fact that they come from different constructors doesn't matter. It's true that `instanceof` checks are all correct, but this is about type-checking arguments to functions which is based on a type's structure, not its constructor (which is what `instanceof` checks).

Right now in TypeScript, an instance of _any_ variant of _any_ type can be passed to an argument of type `option.Option`, because they are _all_ structurally equivalent with `option.None`. This really undermines type-safety when working with the generated TypeScript definitions.

E.g. the following shouldn't type-check and run in TypeScript, but it does:

```ts
import * as thing from "../thing/thing.mjs";
import * as option from "../gleam_stdlib/gleam/option.mjs";

console.log(`Is thing a none? ${option.is_none(new thing.Thing())}`); // Prints false
console.log(`Is thing a some? ${option.is_some(new thing.Thing())}`); // Prints true
```

The relevant docs are here: https://www.typescriptlang.org/docs/handbook/type-compatibility.html.

Reply:
Those are different classes with differing constructors and as such are not structurally equivalent. Specifically the prototype property differs.

Reply:
Unfortunately the prototype isn't considered as part of structural typing in TypeScript! If it was then indeed this issue wouldn't exist. See [this TS playground](https://www.typescriptlang.org/play/?#code/MYGwhgzhAEDCYBdoG8BQ1oDswFsCm0AvNAEQBmIArmWQJ4moC+qqokMAIgPYDmK6WXAWLkAlgBMuDZqjKVMwBKK6ZowABZ5gAa3gIAFMEQAuOIgCU-DMBUQuIPADoQvfSSMJH2fEVLQA1GqIXkLmANwCNph2Ds6u7ojQotEIYAp4XGRmSCIB0IaJyRCp6ZnZ5uGRtvZOLjxuHkkpacAZWdx8uYEFSEUlrWUdFREyGlq6iPqYeADu2frDrJo6elOz0B0LlUA) for a simple example. I was surprised by this behaviour initially, I suspect many TypeScript devs would be.

To try and simplify things I've opened a separate PR that only tries to address the presence of the `any` type in the generated typings: https://github.com/gleam-lang/gleam/pull/3707. It uses a somewhat different approach to achieve the desired result in that case.

If that PR works out then I'll look into adapting this one, converting to a draft for now.